src/Interpreters/GraceHashJoin.cpp

```diff
@@ -713,11 +713,16 @@ void GraceHashJoin::addBlockToJoinImpl(Block block)
             if (!current_block.rows())
                 return;
         }
-
         auto prev_keys_num = hash_join->getTotalRowCount();
         hash_join->addBlockToJoin(current_block, /* check_limits = */ false);
 
-        if (!hasMemoryOverflow(hash_join))
+        size_t hash_join_total_rows = hash_join->getTotalRowCount();
+        if (hash_join->getJoinedData())
+        {
+            hash_join->getJoinedData()->keys_to_join = hash_join_total_rows;
```

Comment:
this is really confusing code, basically:

```
hash_join->getJoinedData()->keys_to_join = hash_join->getTotalRowCount();
```

do we really need to set `keys_to_join` here and if yes, could you pls refactor it somehow

Reply:
yes, the `keys_to_join` will not be calculated in `HashJoin` as we call `hash_join->addBlockToJoin` in `GraceHashJoin`, because the `checkLimit` is set to `false`. So we need to calculate it separately.

Reply:
so if we anyway will do it after each call to `addBlockToJoin`, could we just pass `check_limits=true`?

Reply:
we can not set `check_limits = true`, because the limits check will be done by `hasMemoryOverflow` of `GraceHashJoin`, which is a soft check, if the `check_limits = true`, it will use a different check (not soft)